source/slang/glsl.meta.slang

```diff
@@ -417,3 +417,1399 @@ ${{{{
+
+// ~GL_KHR_shader_subgroup_shuffle
+
+// GL_KHR_shader_subgroup_shuffle_relative
+
+__generic<T : __BuiltinType>
+public T subgroupShuffleUp(T value, uint delta);
+
+__generic<T : __BuiltinType>
+public T subgroupShuffleDown(T value, uint delta);
+
+__generic<T : __BuiltinType>
+__glsl_extension(GL_KHR_shader_subgroup_shuffle_relative)
+__spirv_version(1.3)
+[ForceInline] T subgroupShuffleUp(T value, uint delta)
+{
+    __target_switch
+    {
+    case glsl:
+        __target_intrinsic "subgroupShuffleUp($0, $1)";
+    case spirv:
+        return spirv_asm {
+            OpCapability GroupNonUniformBallot; 
+            OpGroupNonUniformShuffleUp $$T result Subgroup $value $delta
+        };
+    case hlsl:
+    case cuda:
+    case cpp:
+        // TODO: implementation
+        return value;
+    }
+    return value;
+}
+
+__generic<T : __BuiltinType>
+__glsl_extension(GL_KHR_shader_subgroup_shuffle_relative)
+__spirv_version(1.3)
+[ForceInline] T subgroupShuffleDown(T value, uint delta)
+{
+    __target_switch
+    {
+    case glsl:
+        __target_intrinsic "subgroupShuffleDown($0, $1)";
+    case spirv:
+        return spirv_asm {
+            OpCapability GroupNonUniformBallot; 
+            OpGroupNonUniformShuffleDown $$T result Subgroup $value $delta
+        };
+    case hlsl:
+    case cuda:
+    case cpp:
+        // TODO: implementation
+        return value;
+    }
+    return value;
+}
+
+// ~GL_KHR_shader_subgroup_shuffle_relative
+
+// GL_KHR_shader_subgroup_clustered
+
+__generic<T : __BuiltinArithmeticType>
+public T subgroupClusteredAdd(T value, uint clusterSize);
+
+__generic<T : __BuiltinArithmeticType>
+public T subgroupClusteredMul(T value, uint clusterSize);
+
+__generic<T : __BuiltinArithmeticType>
+public T subgroupClusteredMin(T value, uint clusterSize);
+
+__generic<T : __BuiltinArithmeticType>
+public T subgroupClusteredMax(T value, uint clusterSize);
+
+__generic<T : __BuiltinIntegerType>
+public T subgroupClusteredAnd(T value, uint clusterSize);
+
+__generic<T : __BuiltinIntegerType>
+public T subgroupClusteredOr(T value, uint clusterSize);
+
+__generic<T : __BuiltinIntegerType>
+public T subgroupClusteredXor(T value, uint clusterSize);
+
+__generic<T : __BuiltinArithmeticType>
+__glsl_extension(GL_KHR_shader_subgroup_clustered)
+__spirv_version(1.3)
+    [ForceInline] T subgroupClusteredAdd(T value, uint clusterSize)
+{
+    __target_switch
+    {
+    case glsl:
+        __target_intrinsic "subgroupClusteredAdd($0, $1)";
+    case spirv:
+        if (__isFloat<T>())
+            return spirv_asm {OpCapability GroupNonUniformArithmetic; OpCapability GroupNonUniformClustered; 
+            OpGroupNonUniformFAdd $$T result Subgroup ClusteredReduce $value $clusterSize};
+        else if (__isSignedInt<T>())
+        {
+            return spirv_asm
+            {
+                OpCapability GroupNonUniformArithmetic; OpCapability GroupNonUniformClustered;
+                // TODO: use the correct integer width
```

Comment:
I don't think you can use a comment notation inside of spirv_asm.
Or at least we don't do it as far as I know.

Reply:
this was code based off of the hlsl.meta, it works fine in tests